include/deal.II/matrix_free/fe_remote_evaluation.h

```diff
@@ -0,0 +1,1085 @@
+  {};
+
+
+
+  /**
+   * Check if given object is FEPointEvaluation object.
+   */
+  template <typename T>
+  struct is_FEPointEvaluation : std::false_type
+  {};
+
+  template <int n_components, int dim, int spacedim, typename Number>
+  struct is_FEPointEvaluation<
+    dealii::FEPointEvaluation<n_components, dim, spacedim, Number>>
+    : std::true_type
+  {};
+
+
+
+  /**
+   * Type traits for supported FEEvaluationTypes. Different FEEvaluationTypes
+   * need different communication objects and different access to data at
+   * quadrature points. Each specialization defines its CommunicationObjectType
+   * and if a two level CRS structure is needed to access the data by the
+   * memeber `cell_face_pairs`. The same type with a different numer of components 
+   * can be obtained with FEEvaluationTypeComponents.
+   */
+  template <typename FEEvaluationType, bool is_face>
+  struct FEEvaluationTypeTraits
+  {};
+
+
+
+  /**
+   * Specialization for FEEvaluation.
+   */
+  template <int dim,
+            int fe_degree,
+            int n_q_points_1d,
+            int n_components_,
+            typename Number,
+            typename VectorizedArrayType>
+  struct FEEvaluationTypeTraits<FEEvaluation<dim,
+                                             fe_degree,
+                                             n_q_points_1d,
+                                             n_components_,
+                                             Number,
+                                             VectorizedArrayType>,
+                                false>
+  {
+    static constexpr bool cell_face_pairs = false;
+
+    template <int n_components>
+    using FEEvaluationTypeComponents = FEEvaluation<dim,
+                                                    fe_degree,
+                                                    n_q_points_1d,
+                                                    n_components,
+                                                    Number,
+                                                    VectorizedArrayType>;
+
+    using CommunicationObjectType = std::vector<
+      std::pair<std::shared_ptr<Utilities::MPI::RemotePointEvaluation<dim>>,
+                std::vector<std::pair<unsigned int, unsigned int>>>>;
+  };
+
+
+
+  /**
+   * Specialization for FEFaceEvaluation.
+   */
+  template <int dim,
+            int fe_degree,
+            int n_q_points_1d,
+            int n_components_,
+            typename Number,
+            typename VectorizedArrayType>
+  struct FEEvaluationTypeTraits<FEFaceEvaluation<dim,
+                                                 fe_degree,
+                                                 n_q_points_1d,
+                                                 n_components_,
+                                                 Number,
+                                                 VectorizedArrayType>,
+                                true>
+  {
+    static constexpr bool cell_face_pairs = false;
```

Comment:
Shouldn't this be true!?

Reply:
No. This is because `FEFaceEvaluation::reinit(face_index)` takes a unique face batch ID instead of a `cell_index` and `face_number`. This is contrary to `FEPointEvaluation::reinit(cell_index,face_number)` (if setup for faces).